core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,360 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+import io.trino.execution.resourcegroups.IndexedPriorityQueue;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicLongArray;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.lang.Double.isNaN;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Help in finding the skewness across writers when writing partitioned data using preferred partitioning.
+ * It then tries to uniformly distribute the biggest partitions from skewed writers to all the available writers.
+ * <p>
+ * Example:
+ * <p>
+ * Before: For three writers with skewed partitions
+ *      Writer 1 -> No partition assigned -> 0 bytes
+ *      Writer 2 -> No partition assigned -> 0 bytes
+ *      Writer 3 -> Partition 1 (100MB) + Partition 2 (100MB) + Partition 3 (100MB) ->  300 MB
+ * <p>
+ * After scaling:
+ *      Writer 1 -> Partition 1 (50MB) + Partition 3 (50MB) -> 100 MB
+ *      Writer 2 -> Partition 2 (50MB) -> 50 MB
+ *      Writer 3 -> Partition 1 (150MB) + Partition 2 (150MB) + Partition 3 (150MB) ->  450 MB
+ */
+@ThreadSafe
+public class UniformPartitionRebalancer
+{
+    private static final Logger log = Logger.get(UniformPartitionRebalancer.class);
+    private static final long MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES = DataSize.of(100, MEGABYTE).toBytes();
+
+    // If the percentage difference between the two writers with maximum and minimum physical written bytes
+    // since last rebalance is above 0.7 (or 70%), then we consider them skewed.
+    private static final double SKEWNESS_THRESHOLD = 0.7;
+
+    private final List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers;
+    private final long writerMinSize;
+    private final int numberOfWriters;
+
+    private final AtomicReference<List<List<Integer>>> partitionToWritersAssignment;
```

Comment:
since content of `AtomicReference` doesn't change, you can just keep it as:
```
List<Integer>[] partitionToWritersAssignment;
```
nit: consider adding a comment that it's thread safe because `CopyOnWriteArrayList` are used